src/dune_rules/install_entry.ml

```diff
@@ -45,13 +69,41 @@ module File = struct
     let to_file_bindings_unexpanded t ~expand_str ~dir ~dune_syntax =
       match t with
       | File_binding file_binding -> Memo.return [ file_binding ]
-      | Glob_files glob_files ->
+      | Glob_files { glob_files; prefix } ->
         let open Memo.O in
-        let+ paths = Glob_files_expand.memo glob_files ~f:expand_str ~base_dir:dir in
+        let* glob_expanded =
+          Glob_files_expand.memo glob_files ~f:expand_str ~base_dir:dir
+        in
         let glob_loc = String_with_vars.loc glob_files.glob in
-        List.map paths ~f:(fun path ->
+        let glob_prefix = Glob_files_expand.Expanded.prefix glob_expanded in
+        let+ prefix_loc_opt =
+          Memo.Option.map prefix ~f:(fun prefix_sw ->
+            let+ prefix = expand_str prefix_sw in
+            prefix, String_with_vars.loc prefix_sw)
+        in
+        List.map (Glob_files_expand.Expanded.matches glob_expanded) ~f:(fun path ->
           let src = glob_loc, path in
-          File_binding.Unexpanded.make ~src ~dst:src ~dune_syntax)
+          let dst =
+            match prefix_loc_opt with
+            | None -> src
+            | Some (prefix, prefix_loc) ->
+              let path_without_prefix =
+                match String.drop_prefix path ~prefix:glob_prefix with
+                | Some s -> s
+                | None ->
+                  Code_error.raise
+                    ~loc:glob_loc
+                    (Printf.sprintf
+                       "This glob has a prefix %s, but the match %s does not begin with \
```

Comment:
In `Code_error.raise`, the first argument should be the same for all errors. This is done so that it's easy to grep for a specific type of error just by the name. The parameters of the error should be included in the second argument. So for example:

```ocaml
Code_error.raise "Glob has a prefix which does not match the prefix of the match"
  [ "glob_prefix", ..
  ; "match_prefix", ..
  ]
```

Reply:
fixed